tests/test_bookmark.py

```diff
@@ -0,0 +1,162 @@
+
+    def test_run(self):
+        """
+        • Verify for each incremental stream you can do a sync that records bookmarks, and that the format matches expectations.
+        • Verify that the bookmark is the maximum value sent to the target for the replication key.
+        • Verify that a second sync respects the bookmark
+            - All data of the second sync is >= the bookmark from the first sync
+            - The number of records in the 2nd sync is less than the first (This assumes that
+              new data added to the stream is done at a rate slow enough that you haven't
+              doubled the amount of data from the start date to the first sync between
+              the first sync and second sync run in this test)
+        Verify that for the FULL_TABLE stream,
+            • Verify that a bookmark doesn't exist for full table streams.
+            • All data replicated in sync 1 is replicated again in sync 2.
+
+        PREREQUISITE
+        For EACH stream that is incrementally replicated, there are multiple rows of data with
+            different values for the replication key
+        """
+
+        # TDL-20158: Skipping "video_ads", "creatives", "ad_analytics_by_campaign", "ad_analytics_by_creative"
+        # as child bookmark is overridden by parent bookmark
+        streams_to_test = self.expected_streams() - {"video_ads", "creatives", "ad_analytics_by_campaign", "ad_analytics_by_creative"}
+
+        expected_replication_keys = self.expected_replication_keys()
+        expected_replication_methods = self.expected_replication_method()
+
+        ##########################################################################
+        # First Sync
+        ##########################################################################
+        conn_id = connections.ensure_connection(self)
+
+        # Run in check mode
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # table and field selection
+        catalog_entries = [catalog for catalog in found_catalogs
+                           if catalog.get('tap_stream_id') in streams_to_test]
+
+        self.perform_and_verify_table_and_field_selection(
+            conn_id, catalog_entries)
+
+        # Run a first sync job using orchestrator
+        first_sync_record_count = self.run_and_verify_sync(conn_id)
+        first_sync_records = runner.get_records_from_target_output()
+        first_sync_bookmarks = menagerie.get_state(conn_id)
+
+        ##########################################################################
+        # Update State Between Syncs
+        ##########################################################################
+
+        new_states = {'bookmarks': dict()}
+        simulated_states = self.calculated_states_by_stream(
+            first_sync_bookmarks)
+        for stream, new_state in simulated_states.items():
+            new_states['bookmarks'][stream] = new_state
```

Comment:
If the `calculated_states_by_stream()` returns the state in a dict format with stream name as keys, then I don't see any use of this for loop? If so can we please remove it?

Reply:
Updated the code to directly assign dictionary, instead of looping through each stream.